src/goto-symex/goto_symex_state.cpp

```diff
@@ -455,8 +454,9 @@ bool goto_symex_statet::l2_thread_read_encoding(
       symex_renaming_levelt::increase_counter(level2_it);
       a_s_read.first=level2.current_count(l1_identifier);
     }
+    const renamedt<ssa_exprt, L2> l2_false_case = set_l2_indices(ssa_l1, ns);
 
-    to_ssa_expr(tmp.false_case()).set_level_2(a_s_read.first);
+    if_exprt tmp{cond.as_expr(), l2_true_case.get(), l2_false_case.get()};
```

Comment:
Maybe we can delay the construction even further and just use `l2_true_case`, `l2_false_case` in the branches below? Overall your changes here indeed make the code much nicer!